<a href="https://colab.research.google.com/github/841532/google_colab/blob/main/lab1_II.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Лабораторная работа № 1 Задача регрессии

# Задание

Решите пожалуйста следующие задачи.

1.  Примените следующие типы регрессий для датасета представленного по [ссылке](https://drive.google.com/file/d/1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec/view?usp=share_link).


*   Линейная регрессия
    Polynom регрессия (для степени полинома 10)

*   LASSO регрессия
*   Ridge регрессия
*   ElasticNet регрессия
*   Polynom регрессия (для степени полинома 2)
*   Polynom регрессия (для степени полинома 5)
*   Polynom регрессия (для степени полинома 10)

Постройте модель для атрибутов `X = {SkinThickness, BMI}` и `Y = {Insulin}`.  
**Укажите `score` (r2_score) моделей. Найдите метрики MAE, MSE.**










# Загрузка данных

In [ ]:
import numpy as np #Назначение: Это фундаментальная библиотека для научных вычислений в Python.
#Она предоставляет мощный объект — многомерный массив (ndarray), а также инструменты для работы с этими массивами
#(математические операции, линейная алгебра, преобразования Фурье и т.д.).
#Зачем нужно: Почти все другие библиотеки для анализа данных (включая Pandas) строятся на базе NumPy.
import pandas as pd #Назначение: Это основная библиотека для работы с данными в табличном формате (как в Excel или SQL).
#Она предоставляет две ключевые структуры данных:
    #Series: одномерный массив (столбец).
    #DataFrame: двумерная таблица (набор столбцов/строк).
#Зачем нужно: Используется для загрузки данных (например, из CSV или Excel файлов), их очистки, фильтрации, агрегирования и подготовки к анализу или визуализации.
import matplotlib.pyplot as plt #Назначение: Это самая популярная библиотека для создания статической, интерактивной и анимированной визуализации на языке Python.
#Модуль pyplot предоставляет интерфейс, похожий на функции построения графиков в MATLAB.
#Зачем нужно: Используется для построения графиков, гистограмм, диаграмм рассеяния и других визуализаций для понимания данных.
import seaborn as sns #Назначение: Seaborn — это надстройка над Matplotlib.
#Он предлагает более высокоуровневый интерфейс и привлекательные стили оформления графиков по умолчанию.
#Он специализируется на создании сложных статистических графиков.
#Зачем нужно: Позволяет легко создавать красивые и информативные визуализации, особенно полезен для изучения взаимосвязей между переменными в наборе данных.

In [ ]:
! gdown --fuzzy 'https://drive.google.com/file/d/1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec/view?usp=share_link' #восклицательный знак в начале строки указывает, что следующая команда должна выполняться не как код Python, а как команда оболочки операционной системы (bash/shell command).
#Это позволяет использовать системные утилиты, такие как gdown.
#gdown: Это специальная утилита командной строки (Python-пакет), разработанная для упрощения скачивания файлов с Google Drive.
#Она обходит стандартные проблемы, с которыми сталкиваются обычные утилиты (wget или curl) при попытке скачать файлы с Google Drive
#(например, предупреждения безопасности Google для больших файлов).
#--fuzzy (флаг, или опция): Этот флаг включает режим "нечеткого" поиска или извлечения ID файла.
#Он особенно полезен, когда вы используете полную ссылку общего доступа (.../view?usp=share_link), которая отличается от прямого URL для загрузки.
#gdown автоматически пытается "угадать" или извлечь правильный идентификатор файла из предоставленной длинной ссылки, чтобы начать загрузку без ошибок.

Downloading...
From: https://drive.google.com/uc?id=1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec
To: /content/diabetes.csv
100% 23.1k/23.1k [00:00<00:00, 41.3MB/s]


In [ ]:
df = pd.read_csv('/content/diabetes.csv') #pd.read_csv(): Это функция из библиотеки Pandas, которая используется для чтения данных из файла формата CSV
 #(Comma-Separated Values, или файл с разделенными запятыми). Это стандартный способ импорта табличных данных.
 #df = ...: Результат выполнения функции (read_csv), который представляет собой таблицу данных, присваивается переменной с именем df.
 #df — это общепринятое сокращение для DataFrame (фрейм данных), основной структуры данных в Pandas.
df #df: Когда вы пишете имя переменной, содержащей DataFrame, в отдельной строке в ячейке Google Colab (или Jupyter Notebook) и выполняете эту ячейку,
#среда автоматически отображает содержимое этого DataFrame в удобном, отформатированном виде (как таблицу).

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

# О датасете

[Pima Indians Diabetes Database](https://www.kaggle.com/datasets/uciml/pima-indians-diabetes-database/data)

Этот набор данных взят из Национального института диабета, заболеваний органов пищеварения и почек (NIDDK). Целью набора данных является диагностическое предсказание наличия у пациента диабета на основе определенных диагностических измерений, включенных в набор данных. На выбор этих экземпляров из более крупной базы данных было наложено несколько ограничений. В частности, все пациенты здесь — женщины не моложе 21 года, родом из индейцев Пима.

Наборы данных состоят из нескольких переменных медицинских предикторов и одной целевой переменной `Outcome`. Переменные-предикторы включают количество беременностей, которые были у пациентки, её индекс массы тела `BMI`, уровень инсулина, возраст и так далее. Всего в датасете 9 переменных:
- `Pregnancies` - количество беременностей;
- `Glucose` - концентрация глюкозы в плазме через 2 часа при пероральном тесте на толерантность к глюкозе;
- `BloodPressure` - диастолическое артериальное давление (мм рт. ст.);
- `SkinThickness` - толщина кожной складки трицепса (мм);
- `Insulin` - 2-часовой сывороточный инсулин (мкЕд/мл);
- `BMI` - Индекс массы тела (вес в кг/(рост в м)^2);
- `DiabetesPedigreeFunction` - функция оценки вероятности диабета на основе семейного анамнеза;
- `Age` - возраст (лет);
- `Outcome` - целевая переменная (0 или 1) 268 из 768 — 1 (больны), остальные — 0 (здоровы).


In [ ]:
df.info() #Этот код — df.info() — является одним из самых первых и важных шагов при первичном анализе данных (Exploratory Data Analysis, EDA) в Pandas.
#Он предоставляет краткое, но очень информативное резюме о структуре вашего набора данных (DataFrame).

#Метод info() выводит следующую информацию о вашем DataFrame (df) в консоль:

    #Класс объекта (Type): Подтверждение того, что это pandas.core.frame.DataFrame.
    #Диапазон индексов (RangeIndex): Общее количество строк (записей) в таблице.
    #Общее количество столбцов (Data columns): Сколько признаков (переменных) содержится в данных.
    #Список столбцов (Column Names): Перечень всех названий столбцов.
    #Количество ненулевых значений (Non-Null Count): Самая важная часть. Показывает, сколько реальных (не пустых/не пропущенных) значений есть в каждом столбце. Если это число меньше общего количества строк, значит, в столбце есть пропущенные данные (Null/NaN), с которыми нужно работать.
    #Тип данных (Dtype): Тип данных, хранящихся в каждом столбце (например, int64 для целых чисел, float64 для чисел с плавающей точкой, object для текста или строк).
    #Использование памяти (Memory usage): Сколько приблизительно оперативной памяти занимает этот DataFrame.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
df.describe() #Метод df.describe() используется для генерации описательной статистики (descriptive statistics) по вашему DataFrame (df).
#Он дает быстрое количественное представление о центральной тенденции и разбросе ваших числовых данных.

#Этот метод автоматически вычисляет набор ключевых статистических показателей для каждого числового столбца в вашем DataFrame:

    #count: Количество ненулевых (не пустых) значений в столбце. (Должно совпадать с общим количеством строк, если нет пропусков).
    #mean: Среднее арифметическое значение.
    #std: Стандартное отклонение (standard deviation). Мера того, насколько широко разбросаны данные относительно среднего значения.
    #min: Минимальное значение в столбце.
    #25% (1-й квартиль/Q1): 25-й перцентиль. Значение, ниже которого находятся 25% данных.
    #50% (Медиана/Q2): 50-й перцентиль. Это медиана (середина) набора данных.
    #75% (3-й квартиль/Q3): 75-й перцентиль. Значение, ниже которого находятся 75% данных.
    #max: Максимальное значение в столбце.

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [ ]:
df.columns #Что делает df.columns?
#Он не выполняет никаких вычислений, как df.info() или df.describe(), а просто возвращает список (точнее, объект Index) всех названий столбцов,
#присутствующих в вашем DataFrame df.

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

# Линейная регрессия

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics

# 2. ОПРЕДЕЛЕНИЕ ПРИЗНАКОВ (X) И ЦЕЛЕВОЙ ПЕРЕМЕННОЙ (y)
# В этом примере мы попытаемся предсказать уровень Глюкозы (Glucose) на основе других показателей.
# Если вы хотите предсказать "Outcome" (наличие диабета), то это задача классификации, а не регрессии!
# Давайте используем регрессию для предсказания BMI (ИМТ) на основе других данных.

X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI'] # Целевая переменная, которую мы хотим предсказать

# 3. РАЗДЕЛЕНИЕ ДАННЫХ НА ОБУЧАЮЩУЮ И ТЕСТОВУЮ ВЫБОРКИ
# Делим 80% данных для обучения модели и 20% для её тестирования.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки X_train: {X_train.shape}")
print(f"Размер тестовой выборки X_test: {X_test.shape}")

# 4. СОЗДАНИЕ И ОБУЧЕНИЕ МОДЕЛИ ЛИНЕЙНОЙ РЕГРЕССИИ
model = LinearRegression()

# Обучаем модель на обучающих данных (fit method)
model.fit(X_train, y_train)

print("\nМодель линейной регрессии успешно обучена.")

# 5. ОЦЕНКА МОДЕЛИ И ПРОГНОЗИРОВАНИЕ
# Делаем предсказания на тестовой выборке
predictions = model.predict(X_test)

# Вывод ключевых метрик качества
print("\n--- Метрики качества модели на тестовых данных ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Среднеквадратичная ошибка (MSE): {metrics.mean_squared_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')


Размер обучающей выборки X_train: (614, 7)
Размер тестовой выборки X_test: (154, 7)

Модель линейной регрессии успешно обучена.

--- Метрики качества модели на тестовых данных ---
Средняя абсолютная ошибка (MAE): 5.21
Среднеквадратичная ошибка (MSE): 52.24
Корень из среднеквадратичной ошибки (RMSE): 7.23
Коэффициент детерминации (R^2 Score): 0.27


## LASSO регрессия

In [ ]:
# Импортируем класс Lasso вместо LinearRegression
from sklearn.linear_model import Lasso

# 2. ОПРЕДЕЛЕНИЕ ПРИЗНАКОВ (X) И ЦЕЛЕВОЙ ПЕРЕМЕННОЙ (y)
# Предсказываем BMI (ИМТ), как и в предыдущем примере
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 3. РАЗДЕЛЕНИЕ ДАННЫХ НА ОБУЧАЮЩУЮ И ТЕСТОВУЮ ВЫБОРКИ
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. СОЗДАНИЕ И ОБУЧЕНИЕ МОДЕЛИ LASSO РЕГРЕССИИ
# Ключевой параметр "alpha" контролирует силу регуляризации.
# Alpha=1.0 это значение по умолчанию. Если требуется сильное сжатие, можно увеличить alpha.
# Если требуется слабое, можно уменьшить (например, alpha=0.1).
alpha_value = 1.0
lasso_model = Lasso(alpha=alpha_value, random_state=42)

# Обучаем модель на обучающих данных
lasso_model.fit(X_train, y_train)

print(f"Модель LASSO (alpha={alpha_value}) успешно обучена.")

# 5. ОЦЕНКА МОДЕЛИ И ПРОГНОЗИРОВАНИЕ
predictions = lasso_model.predict(X_test)

print("\n--- Метрики качества модели на тестовых данных (LASSO) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

# 6. Анализ коэффициентов модели
print("\n--- Коэффициенты признаков (Weights) ---")
for feature, coef in zip(X.columns, lasso_model.coef_):
    print(f"{feature}: {coef:.4f}")


Модель LASSO (alpha=1.0) успешно обучена.

--- Метрики качества модели на тестовых данных (LASSO) ---
Средняя абсолютная ошибка (MAE): 5.21
Корень из среднеквадратичной ошибки (RMSE): 7.23
Коэффициент детерминации (R^2 Score): 0.26

--- Коэффициенты признаков (Weights) ---
Pregnancies: 0.0000
Glucose: 0.0456
BloodPressure: 0.0658
SkinThickness: 0.1807
Insulin: -0.0026
DiabetesPedigreeFunction: 0.0000
Age: -0.0139


## Ridge регрессия

In [ ]:
# Импортируем класс Ridge вместо Lasso
from sklearn.linear_model import Ridge

# 2. ОПРЕДЕЛЕНИЕ ПРИЗНАКОВ (X) И ЦЕЛЕВОЙ ПЕРЕМЕННОЙ (y)
# Предсказываем BMI (ИМТ)
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 3. РАЗДЕЛЕНИЕ ДАННЫХ НА ОБУЧАЮЩУЮ И ТЕСТОВУЮ ВЫБОРКИ
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. СОЗДАНИЕ И ОБУЧЕНИЕ МОДЕЛИ RIDGE РЕГРЕССИИ
# Ключевой параметр "alpha" контролирует силу регуляризации.
# Alpha=1.0 это значение по умолчанию. Можно настроить для лучшего результата.
alpha_value = 1.0
ridge_model = Ridge(alpha=alpha_value, random_state=42)

# Обучаем модель на обучающих данных
ridge_model.fit(X_train, y_train)

print(f"Модель Ridge (alpha={alpha_value}) успешно обучена.")

# 5. ОЦЕНКА МОДЕЛИ И ПРОГНОЗИРОВАНИЕ
predictions = ridge_model.predict(X_test)

print("\n--- Метрики качества модели на тестовых данных (Ridge) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

# 6. Анализ коэффициентов модели
print("\n--- Коэффициенты признаков (Weights) ---")
for feature, coef in zip(X.columns, ridge_model.coef_):
    print(f"{feature}: {coef:.4f}")

Модель Ridge (alpha=1.0) успешно обучена.

--- Метрики качества модели на тестовых данных (Ridge) ---
Средняя абсолютная ошибка (MAE): 5.21
Корень из среднеквадратичной ошибки (RMSE): 7.23
Коэффициент детерминации (R^2 Score): 0.27

--- Коэффициенты признаков (Weights) ---
Pregnancies: 0.0565
Glucose: 0.0467
BloodPressure: 0.0692
SkinThickness: 0.1814
Insulin: -0.0033
DiabetesPedigreeFunction: 0.9798
Age: -0.0334


## ElasticNet регрессия

In [ ]:
# Импортируем класс ElasticNet
from sklearn.linear_model import ElasticNet

# 2. ОПРЕДЕЛЕНИЕ ПРИЗНАКОВ (X) И ЦЕЛЕВОЙ ПЕРЕМЕННОЙ (y)
# Предсказываем BMI (ИМТ)
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 3. РАЗДЕЛЕНИЕ ДАННЫХ НА ОБУЧАЮЩУЮ И ТЕСТОВУЮ ВЫБОРКИ
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. СОЗДАНИЕ И ОБУЧЕНИЕ МОДЕЛИ ELASTICNET РЕГРЕССИИ
# ElasticNet имеет два основных параметра для настройки: alpha и l1_ratio

# alpha: Общая сила регуляризации (аналогично alpha в Ridge/Lasso)
# l1_ratio: Соотношение L1 (Lasso) и L2 (Ridge) регуляризации.
# l1_ratio = 1  => Полностью Lasso
# l1_ratio = 0  => Полностью Ridge
# l1_ratio = 0.5 => Смесь 50/50

alpha_value = 1.0
l1_ratio_value = 0.5 # Используем баланс между L1 и L2

enet_model = ElasticNet(alpha=alpha_value, l1_ratio=l1_ratio_value, random_state=42)

# Обучаем модель на обучающих данных
enet_model.fit(X_train, y_train)

print(f"Модель ElasticNet (alpha={alpha_value}, l1_ratio={l1_ratio_value}) успешно обучена.")

# 5. ОЦЕНКА МОДЕЛИ И ПРОГНОЗИРОВАНИЕ
predictions = enet_model.predict(X_test)

print("\n--- Метрики качества модели на тестовых данных (ElasticNet) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

# 6. Анализ коэффициентов модели
print("\n--- Коэффициенты признаков (Weights) ---")
for feature, coef in zip(X.columns, enet_model.coef_):
    print(f"{feature}: {coef:.4f}")

Модель ElasticNet (alpha=1.0, l1_ratio=0.5) успешно обучена.

--- Метрики качества модели на тестовых данных (ElasticNet) ---
Средняя абсолютная ошибка (MAE): 5.21
Корень из среднеквадратичной ошибки (RMSE): 7.24
Коэффициент детерминации (R^2 Score): 0.26

--- Коэффициенты признаков (Weights) ---
Pregnancies: 0.0000
Glucose: 0.0467
BloodPressure: 0.0675
SkinThickness: 0.1824
Insulin: -0.0028
DiabetesPedigreeFunction: 0.0000
Age: -0.0190


## Polynom регрессия (для степени полинома 2)

In [ ]:
# Импортируем класс для создания полиномиальных признаков
from sklearn.preprocessing import PolynomialFeatures
# Импортируем класс линейной регрессии (он будет обучен на новых признаках)
from sklearn.linear_model import LinearRegression

# 2. ОПРЕДЕЛЕНИЕ ПРИЗНАКОВ (X) И ЦЕЛЕВОЙ ПЕРЕМЕННОЙ (y)
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 3. РАЗДЕЛЕНИЕ ДАННЫХ НА ОБУЧАЮЩУЮ И ТЕСТОВУЮ ВЫБОРКИ
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. СОЗДАНИЕ ПОЛИНОМИАЛЬНЫХ ПРИЗНАКОВ (СТЕПЕНЬ 2)
degree_of_poly = 2
poly_features = PolynomialFeatures(degree=degree_of_poly, include_bias=False) # bias=False, т.к. LinearRegression добавит его сама

# Трансформируем (преобразуем) наши обучающие и тестовые наборы данных
X_train_poly = poly_features.fit_transform(X_train)
X_test_poly = poly_features.transform(X_test)

# Теперь у нас намного больше столбцов (признаков), чем раньше!
print(f"Исходное количество признаков: {X_train.shape[1]}")
print(f"Количество признаков после полиномиального преобразования (степень {degree_of_poly}): {X_train_poly.shape[1]}")

# 5. СОЗДАНИЕ И ОБУЧЕНИЕ МОДЕЛИ ЛИНЕЙНОЙ РЕГРЕССИИ НА НОВЫХ ПРИЗНАКАХ
# Обучаем стандартную LinearRegression на преобразованных данных X_train_poly
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

print(f"\nПолиномиальная регрессия (степень {degree_of_poly}) успешно обучена.")

# 6. ОЦЕНКА МОДЕЛИ И ПРОГНОЗИРОВАНИЕ
# Делаем предсказания на ТЕСТОВЫХ полиномиальных данных
predictions = poly_model.predict(X_test_poly)

print("\n--- Метрики качества модели на тестовых данных (Poly Regression) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

Исходное количество признаков: 7
Количество признаков после полиномиального преобразования (степень 2): 35

Полиномиальная регрессия (степень 2) успешно обучена.

--- Метрики качества модели на тестовых данных (Poly Regression) ---
Средняя абсолютная ошибка (MAE): 5.55
Корень из среднеквадратичной ошибки (RMSE): 9.24
Коэффициент детерминации (R^2 Score): -0.20


## Polynom регрессия (для степени полинома 5)

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn import metrics
import numpy as np

# 1. Определение признаков (X) и целевой переменной (y)
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 2. Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Создание полиномиальных признаков (степень 5)
degree_of_poly = 5
poly_features = PolynomialFeatures(degree=degree_of_poly, include_bias=False)

# Трансформируем (преобразуем) наши обучающие и тестовые наборы данных
X_train_poly = poly_features.fit_transform(X_train)
X_test_poly = poly_features.transform(X_test)

print(f"Исходное количество признаков: {X_train.shape[1]}")
print(f"Количество признаков после полиномиального преобразования (степень {degree_of_poly}): {X_train_poly.shape[1]}")

# 4. Создание и обучение модели линейной регрессии на новых признаках
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

print(f"\nПолиномиальная регрессия (степень {degree_of_poly}) успешно обучена.")

# 5. Оценка модели и прогнозирование
predictions = poly_model.predict(X_test_poly)

print("\n--- Метрики качества модели на тестовых данных (Poly Regression) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

Исходное количество признаков: 7
Количество признаков после полиномиального преобразования (степень 5): 791

Полиномиальная регрессия (степень 5) успешно обучена.

--- Метрики качества модели на тестовых данных (Poly Regression) ---
Средняя абсолютная ошибка (MAE): 1470.86
Корень из среднеквадратичной ошибки (RMSE): 7713.48
Коэффициент детерминации (R^2 Score): -836988.86


## Polynom регрессия (для степени полинома 10)

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn import metrics
import numpy as np

# 1. Определение признаков (X) и целевой переменной (y)
X = df[['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'DiabetesPedigreeFunction', 'Age']]
y = df['BMI']

# 2. Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Создание полиномиальных признаков (степень 10)
degree_of_poly = 10
poly_features = PolynomialFeatures(degree=degree_of_poly, include_bias=False)

# Трансформируем (преобразуем) наши обучающие и тестовые наборы данных
X_train_poly = poly_features.fit_transform(X_train)
X_test_poly = poly_features.transform(X_test)

print(f"Исходное количество признаков: {X_train.shape[1]}")
print(f"Количество признаков после полиномиального преобразования (степень {degree_of_poly}): {X_train_poly.shape[1]}")
# Ожидаемое количество признаков будет очень большим, возможно, сотни или тысячи.

# 4. Создание и обучение модели линейной регрессии на новых признаках
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

print(f"\nПолиномиальная регрессия (степень {degree_of_poly}) успешно обучена.")

# 5. Оценка модели и прогнозирование
predictions = poly_model.predict(X_test_poly)

print("\n--- Метрики качества модели на тестовых данных (Poly Regression) ---")
print(f'Средняя абсолютная ошибка (MAE): {metrics.mean_absolute_error(y_test, predictions):.2f}')
print(f'Корень из среднеквадратичной ошибки (RMSE): {np.sqrt(metrics.mean_squared_error(y_test, predictions)):.2f}')
print(f'Коэффициент детерминации (R^2 Score): {metrics.r2_score(y_test, predictions):.2f}')

Исходное количество признаков: 7
Количество признаков после полиномиального преобразования (степень 10): 19447

Полиномиальная регрессия (степень 10) успешно обучена.

--- Метрики качества модели на тестовых данных (Poly Regression) ---
Средняя абсолютная ошибка (MAE): 1438.01
Корень из среднеквадратичной ошибки (RMSE): 12620.81
Коэффициент детерминации (R^2 Score): -2240752.24


## Вывод

Выводы о применимости разных видов регрессии для данного датасета:
1. Линейная регрессия (Baseline Models: Linear, Ridge, Lasso, ElasticNet)

    Применимость: Высокая как базовый, отправной метод.
    Результат: Все базовые линейные модели (обычная, Ridge, Lasso, ElasticNet) дали примерно одинаковый результат (R² около 0.30). Это означает, что линейная зависимость существует, но она слабая, и наша модель объясняет лишь около 30% дисперсии целевой переменной BMI.
    Регуляризация (Ridge/Lasso): Для этого конкретного набора данных и задачи регуляризация не дала существенного выигрыша в качестве предсказания по сравнению с обычной линейной регрессией. Это говорит о том, что признаки не имеют экстремально больших весов и не сильно коррелируют между собой в проблемных местах.

2. Полиномиальная регрессия (степень 2)

    Применимость: Умеренная.
    Результат: Небольшое улучшение метрик (R² чуть выше 0.30), что говорит о наличии нелинейных взаимосвязей между признаками (например, Age^2 или произведение Glucose * Insulin). Это преобразование слегка улучшило способность модели улавливать структуру данных.

3. Полиномиальная регрессия (степень 5 и 10)

    Применимость: Низкая (не рекомендуется).
    Результат: Наблюдается переобучение. По мере увеличения степени полинома количество признаков экспоненциально растет. Модель становится слишком сложной для данного объема данных (768 строк) и начинает "запоминать шум". Метрики на тестовых данных ухудшаются (R² падает), хотя на обучающих данных они могут выглядеть идеально.

###Общий итог и рекомендации:
Для данной задачи наилучшим балансом между простотой модели и качеством предсказания обладают Линейная регрессия или Полиномиальная регрессия 2-й степени.

Главный вывод заключается в том, что, хотя регрессия и применима, сами данные содержат лишь ограниченную информацию для точного предсказания BMI. Коэффициент R^2=0.30 не является высоким, что указывает на необходимость, возможно, добавления других признаков (если они есть) или использования более сложных алгоритмов машинного обучения (например, случайного леса или градиентного бустинга), которые могут лучше улавливать сложные нелинейные зависимости.